### ISSUE TITLE
Support to attach volumes and volumeMounts to the apply and destroy pod. 

### ISSUE THREAD

**Issue by LochanRn** [kubevela/terraform-controller#357]
Ability to support volumes and volumeMount to the terraform Job. 

This allows the user to mount files to the pod using a secret or a configmap which can further be used by the terraform executor pod to while creating the resource. 
For example if the user wants to deploy a vm with ability to pass the cloud-init file to the vm.

**Comment by chivalryq** [comment]
Hi @LochanRn. Are you using terraform-controller to create vm-like cloud resource and encountered the related situations? PR is always welcomed.

**Comment by LochanRn** [comment]
@chivalryq yes I would like to raise a PR for the same and planning to keep the design something as below please let me know if it is fine ?

```
// ConfigurationSpec defines the desired state of Configuration
type ConfigurationSpec struct {
        ...
	VolumeSpec VolumeSpec `json:"volumeSpec,omitempty"`
}

type VolumeSpec struct {
	Volumes []corev1.Volume `json:"volumes,omitempty"`
	VolumeMounts []corev1.VolumeMount `json:"volumeMounts,omitempty"`
}
```

**Comment by chivalryq** [comment]
Do we need to specify the phase(applying/destroying) for volumes to mount? IMO in most of cases we can assume the apply phase need a volume. But is there any chance that we need it in destroying phase? @LochanRn

**Comment by LochanRn** [comment]
So there is a use case where we need to run a shell script to clean up some of the external env. In that case during destroy we would require the volumeMount option during destroy as well :)

**Comment by chivalryq** [comment]
So how about split the spec to 
```
// ConfigurationSpec defines the desired state of Configuration
type ConfigurationSpec struct {
        ...
	Volumes Volumes `json:"volumeSpec,omitempty"`
}

type Volumes struct {
	Apply VolumeSpec
	Destroy VolumeSpec
}

type VolumeSpec struct {
	Volumes []corev1.Volume `json:"volumes,omitempty"`
	VolumeMounts []corev1.VolumeMount `json:"volumeMounts,omitempty"`
}
```

**Comment by LochanRn** [comment]
Hi i was just thinking about the above use-case and was thinking, what if the user wants the volumes in both apply and destroy pod then he will have to add a redundant block twice. @chivalryq  WDYT ?

**Comment by chivalryq** [comment]
I didn't think of that before. But yes it is possible. I found another question, you mentioned that there's a case that a script will be run after destroying. Who are responsible for call that script if we have mount that volumes to destroy job pod. IMO pre-start and post-destroy hooks can be what we need.

**Comment by LochanRn** [comment]
before destroying the vm we run the script and we do that using the provisioner. 

```
        provisioner "local-exec" {
          when    = destroy
          command =  "echo 'Executing pre vm destroy hook'; sleep 5s"
        }
```

**Comment by LochanRn** [comment]
Switching back to this model to avoid redundancy. 
```
// ConfigurationSpec defines the desired state of Configuration
type ConfigurationSpec struct {
        ...
	VolumeSpec VolumeSpec `json:"volumeSpec,omitempty"`
}

type VolumeSpec struct {
	Volumes []corev1.Volume `json:"volumes,omitempty"`
	VolumeMounts []corev1.VolumeMount `json:"volumeMounts,omitempty"`
}
```

**Comment by chivalryq** [comment]
Sure, it's also doable.
